### PR TITLE
Clean C Build

### PR DESCRIPTION
**Pull request by dkfellows**
Squelch the remaining warnings in the FEC C code and turn warnings into errors so we won't let things rot in the future.

### CODE REVIEW

**Comment by dkfellows** [comment]
Squelched a bunch of bad practice (e.g., using `_` as a prefix for globally-visible functions). Instead, we're using `static __attribute__((noinline))` so that the functions concerned have only file-local visibility (and can be named whatever we want) and yet they won't be inlined by an over-zealous compiler. Which is pretty elegant really…

File: c_common/front_end_common_lib/include/recording.h [hunk 1]
```diff
@@ -12,10 +12,10 @@
 #include <spin1_api.h>
 #include <buffered_eieio_defs.h>
 
-#define RECORDING_DMA_COMPLETE_TAG_ID 15
+#define RECORDING_DMA_COMPLETE_TAG_ID	15
```
Comment (rowleya) [inline c_common/front_end_common_lib/include/recording.h]: What is the style issue here?

File: c_common/front_end_common_lib/include/simulation.h [hunk 2]
```diff
@@ -14,38 +14,44 @@
 #include "common-typedefs.h"
 #include <spin1_api.h>
 
-// constant for how many dma ids you can use (caps the values of the tags as
-// well)
-#define MAX_DMA_CALLBACK_TAG 16
+// How many DMA IDs can be used (caps the values of the tags as well)
+#define MAX_DMA_CALLBACK_TAG	16
```
Comment (rowleya) [inline c_common/front_end_common_lib/include/simulation.h]: Is that a tab I see before me?

File: c_common/front_end_common_lib/src/data_specification.c [hunk 3]
```diff
@@ -25,15 +26,15 @@ typedef enum region_elements{
 //!        loaded into the user0 register of the core during the tool chain
 //!        loading.
 //! \return the SDRAM start address for this core.
-address_t data_specification_get_data_address() {
-
+address_t data_specification_get_data_address(void)
+{
```
Comment (rowleya) [inline c_common/front_end_common_lib/src/data_specification.c]: See previous messages re: inline brackets

File: c_common/front_end_common_lib/src/data_specification.c [hunk 4]
```diff
@@ -25,15 +26,15 @@ typedef enum region_elements{
...
     // Get pointer to 1st virtual processor info struct in SRAM
     vcpu_t *sark_virtual_processor_info = (vcpu_t*) SV_VCPU;
 
     // Get the address this core's DTCM data starts at from the user data member
     // of the structure associated with this virtual processor
-    address_t address =
-        (address_t) sark_virtual_processor_info[spin1_get_core_id()].user0;
+    address_t address = (address_t)
+	    sark_virtual_processor_info[spin1_get_core_id()].user0;
```
Comment (rowleya) [inline c_common/front_end_common_lib/src/data_specification.c]: Objection to using tab at start of line...

File: c_common/front_end_common_lib/src/data_specification.c [hunk 5]
```diff
@@ -48,24 +49,24 @@ address_t data_specification_get_data_address() {
...
     }
 
     // Log what we have found
     log_info("magic = %08x, version = %d.%d", address[dse_magic_number],
-             address[dse_version] >> VERSION_SHIFT,
-             address[dse_version] & VERSION_MASK);
-    return (true);
+	    address[dse_version] >> VERSION_SHIFT,
+	    address[dse_version] & VERSION_MASK);
+    return true;
```
Comment (rowleya) [inline c_common/front_end_common_lib/src/data_specification.c]: More objections to tabs...

File: c_common/front_end_common_lib/src/data_specification.c [hunk 6]
```diff
@@ -77,6 +78,7 @@ bool data_specification_read_header(uint32_t* address) {
 //! \return a address_t which represents the absolute SDRAM address for the
 //!         start of the requested region.
 address_t data_specification_get_region(
-        uint32_t region, address_t data_address) {
-    return (address_t) (data_address[REGION_START_INDEX + region]);
+        uint32_t region, address_t data_address)
+{
```
Comment (rowleya) [inline c_common/front_end_common_lib/src/data_specification.c]: Objection to bracket...

File: c_common/front_end_common_lib/src/profiler.c [hunk 7]
```diff
@@ -13,7 +13,9 @@ uint32_t *profiler_output = NULL;
 //---------------------------------------
 // Functions
 //---------------------------------------
-void profiler_init(uint32_t* data_region) {
+void profiler_init(
+	uint32_t* data_region)
+{
```
Comment (rowleya) [inline c_common/front_end_common_lib/src/profiler.c]: There's that bracket again...

File: c_common/front_end_common_lib/src/profiler.c [hunk 8]
```diff
@@ -31,10 +33,11 @@ void profiler_init(uint32_t* data_region) {
 }
 
 //---------------------------------------
-void profiler_finalise() {
+void profiler_finalise(void)
+{
```
Comment (rowleya) [inline c_common/front_end_common_lib/src/profiler.c]: Another one...

File: c_common/front_end_common_lib/src/recording.c [hunk 9]
```diff
@@ -185,7 +195,9 @@ static inline void _recording_eieio_packet_handler(
 }
 
 // Work out the space available in the given channel for recording
-static uint32_t compute_available_space_in_channel(uint8_t channel) {
+static NOINLINE uint32_t compute_available_space_in_channel(
```
Comment (rowleya) [inline c_common/front_end_common_lib/src/recording.c]: Is there a good reason to force this to not inline?

File: c_common/front_end_common_lib/src/recording.c [hunk 10]
```diff
@@ -202,22 +214,19 @@ static uint32_t compute_available_space_in_channel(uint8_t channel) {
...
     } else {
-
         // If pointers are equal, buffer is empty if last operation is read
-        return (end_of_buffer_region - buffer_region);
+        return end_of_buffer_region - buffer_region;
     }
 }
 
-static void _recording_write(
+static NOINLINE void write_chunk(
```
Comment (rowleya) [inline c_common/front_end_common_lib/src/recording.c]: As above, is it certain that this shouldn't be inlined?

File: c_common/front_end_common_lib/src/recording.c [hunk 11]
```diff
@@ -290,151 +289,142 @@ static inline bool _recording_write_memory(
...
-        read_request_packet_data *data_ptr, uint n_requests,
-        uint channel, uint8_t *read_pointer, uint32_t space_to_be_read) {
-    data_ptr[n_requests].processor_and_request = 0;
-    data_ptr[n_requests].sequence = 0;
-    data_ptr[n_requests].channel = channel;
-    data_ptr[n_requests].region =
-        g_recording_channels[channel].region_id;
-    data_ptr[n_requests].start_address = (uint32_t) read_pointer;
-    data_ptr[n_requests].space_to_be_read = space_to_be_read;
+static NOINLINE void create_buffer_message(
```
Comment (rowleya) [inline c_common/front_end_common_lib/src/recording.c]: Check why this is forced not inline...

File: c_common/front_end_common_lib/src/recording.c [hunk 12]
```diff
@@ -290,151 +289,142 @@ static inline bool _recording_write_memory(
...
-        msg_size += (n_requests * sizeof(read_request_packet_data));
+        msg_size += n_requests * sizeof(read_request_packet_data);
         msg.length = msg_size;
 
         spin1_send_sdp_msg(&msg, 1);
     }
 }
 
-static void _buffering_in_handler(uint mailbox, uint port) {
+static NOINLINE void buffering_in_handler(
```
Comment (rowleya) [inline c_common/front_end_common_lib/src/recording.c]: Again, check why not inlined...

File: c_common/front_end_common_lib/src/recording.c [hunk 13]
```diff
@@ -488,30 +469,24 @@ bool recording_record(uint8_t channel, void *data, uint32_t size_bytes) {
...
-    for (uint32_t recording_region_id = 0;
-             recording_region_id < n_recording_regions;
-             recording_region_id++) {
-        address_t recording_region_address =
-            region_addresses[recording_region_id];
-        spin1_memcpy(
-             recording_region_address,
-             &g_recording_channels[recording_region_id],
-             sizeof(recording_channel_t));
+static NOINLINE void write_buffer_state_data(void)
```
Comment (rowleya) [inline c_common/front_end_common_lib/src/recording.c]: Check me...

File: c_common/models/live_packet_gather/live_packet_gather.c [hunk 14]
```diff
@@ -228,15 +225,15 @@ void process_incoming_event(uint key) {
...
-            spike_index = buffer_index << 1;
-            buf_pointer[spike_index] = CLAMP16(time);
-            buf_pointer[spike_index + 1] = CLAMP16(time >> 16);
+            if (payload_timestamp) {
+		buf_pointer[spike_index] = CLAMP16(time);
+		buf_pointer[spike_index + 1] = CLAMP16(time >> 16);
+            } else {
+		buf_pointer[spike_index] = 0;
+		buf_pointer[spike_index + 1] = 0;
+            }
```
Comment (rowleya) [inline c_common/models/live_packet_gather/live_packet_gather.c]: TABS!  The spaced version looked better, even on github...

File: c_common/models/live_packet_gather/live_packet_gather.c [hunk 15]
```diff
@@ -252,21 +249,21 @@ void process_incoming_event_payload(uint key, uint payload) {
...
-            buffer_index++;
-        } else if (HAVE_PAYLOAD(packet_type) && payload_timestamp) {
-            buf_pointer[buffer_index] = CLAMP16(time);
+        if (HAVE_PAYLOAD(packet_type)) {
+            if (payload_timestamp) {
+        	buf_pointer[buffer_index] = CLAMP16(time);
+            } else {
+        	buf_pointer[buffer_index] = CLAMP16(
+        		payload >> payload_right_shift);
+            }
```
Comment (rowleya) [inline c_common/models/live_packet_gather/live_packet_gather.c]: For fun, this has mixed spaces and tabs!

File: c_common/models/live_packet_gather/live_packet_gather.c [hunk 16]
```diff
@@ -275,15 +272,15 @@ void process_incoming_event_payload(uint key, uint payload) {
...
             spike_index = buffer_index << 1;
-            buf_pointer[spike_index] = CLAMP16(time);
-            buf_pointer[spike_index + 1] = CLAMP16(time >> 16);
+            if (payload_timestamp) {
+		buf_pointer[spike_index] = CLAMP16(time);
+		buf_pointer[spike_index + 1] = CLAMP16(time >> 16);
+            } else {
+		buf_pointer[spike_index] = CLAMP16(payload);
+		buf_pointer[spike_index + 1] = CLAMP16(payload >> 16);
+            }
```
Comment (rowleya) [inline c_common/models/live_packet_gather/live_packet_gather.c]: Yucky tabs

File: c_common/models/live_packet_gather/live_packet_gather.c [hunk 17]
```diff
@@ -441,18 +436,20 @@ bool configure_sdp_msg(void) {
...
+		    "as payload to each key, not both\n");
+	    return false;
+	}
+	if (!payload_apply_prefix && !HAVE_PAYLOAD(packet_type)) {
+	    log_error(
+		    "Timestamp can either be included as payload prefix or "
+		    "as payload to each key, but current configuration does "
+		    "not specify either of these\n");
+	    return false;
+	}
```
Comment (rowleya) [inline c_common/models/live_packet_gather/live_packet_gather.c]: Tab, tab, tab, tab

**Comment by alan-stokes** [comment]
rowleys comments look valid to me. benched pr till fixed.

File: c_common/front_end_common_lib/include/profiler.h [hunk 18]
```diff
@@ -35,22 +35,28 @@ extern uint32_t *profiler_output;
 //---------------------------------------
 // Inline functions
 //---------------------------------------
-static inline void profiler_write_entry(uint32_t tag) {
+static inline void profiler_write_entry(
+	uint32_t tag)
+{
```
[the remark this answers is not in the document]
Comment (dkfellows) [reply c_common/front_end_common_lib/include/profiler.h]: I think it's easier to read.

[the remark this answers is not in the document]
Comment (rowleya) [reply c_common/front_end_common_lib/include/profiler.h]: Well this is why we have to decide on a style - there are two for brackets, one on the same line and one on the next.  Most of our code has the bracket on the same line, and this is the style we have had for some time... maybe we should update vera to reflect this...

File: c_common/front_end_common_lib/src/recording.c (hunk 12 above)
Comment (dkfellows) [reply c_common/front_end_common_lib/src/recording.c]: This one is used as a callback with `simulation_sdp_callback_on`.

File: c_common/front_end_common_lib/src/data_specification.c (hunk 3 above)
Comment (rowleya) [reply c_common/front_end_common_lib/src/data_specification.c]: OK, but that isn't the style in the rest of the code (not in most of it anyway)...

**Comment by dkfellows** [comment]
#462 effectively supersedes this

**Comment by alan-stokes** [comment]
keeping branch for any reason?

File: c_common/models/common.mk [hunk 19]
```diff
@@ -10,3 +10,7 @@ SOURCE_DIRS += $(SOURCE_DIR)
 APP_OUTPUT_DIR := $(abspath $(CURRENT_DIR)../../spinn_front_end_common/common_model_binaries/)/
 
 include $(CURRENT_DIR)../front_end_common_lib/FrontEndCommon.mk
+
+ifeq ($(GNU),1)
+CFLAGS += -Werror
```
Comment (dkfellows) [inline c_common/models/common.mk:15]: Only bit worth keeping.

Comment (alan-stokes) [reply c_common/models/common.mk:15]: worth a seperate pr then?

**Pull request closed without merging**